pkg/validator/validators.go

```diff
@@ -244,25 +254,149 @@ func (kc *policyValidator) upstreamTrafficSettingValidator(req *admissionv1.Admi
+	log.Debug().Msg("made it to MRC validator")
+	switch req.Operation {
+	case admissionv1.Create:
+		newMRC := &configv1alpha2.MeshRootCertificate{}
+		if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(newMRC); err != nil {
+			return nil, err
+		}
+
+		if newMRC.Spec.TrustDomain == "" {
+			return nil, fmt.Errorf("trustDomain must be non empty for MRC %s/%s", newMRC.GetNamespace(), newMRC.GetName())
+		}
+
+		// if no state is set, MRC will not be considered for rotation
+		if newMRC.Status.State == "" {
+			return nil, nil
+		}
+
+		if cv.activeRootCertificateRotation(newMRC) {
+			return nil, fmt.Errorf("cannot create MRC %s/%s with status %s. Certificate rotation in progress. At most, only 2 MRCs can be in non error or non inactive states",
+				newMRC.GetNamespace(), newMRC.GetName(), newMRC.Status.State)
+		}
+
+		if _, ok := validMRCTransitions[newMRC.Status.State]; !ok {
+			return nil, fmt.Errorf("invalid status %s for MRC %s/%s. Must be %s, %s, %s, %s, %s, %s, or %s",
+				newMRC.Status.State, newMRC.GetNamespace(), newMRC.GetName(), constants.MRCStateValidatingRollout,
+				constants.MRCStateIssuingRollout, constants.MRCStateActive, constants.MRCStateIssuingRollback,
+				constants.MRCStateValidatingRollback, constants.MRCStateInactive, constants.MRCStateError)
+		}
+	case admissionv1.Update:
+		newMRC, oldMRC := &configv1alpha2.MeshRootCertificate{}, &configv1alpha2.MeshRootCertificate{}
+		if err := json.NewDecoder(bytes.NewBuffer(req.Object.Raw)).Decode(newMRC); err != nil {
+			return nil, err
+		}
+		if err := json.NewDecoder(bytes.NewBuffer(req.OldObject.Raw)).Decode(oldMRC); err != nil {
+			return nil, err
+		}
+
+		if !reflect.DeepEqual(oldMRC.Spec.Provider, newMRC.Spec.Provider) {
+			return nil, fmt.Errorf("cannot update certificate provider settings for MRC %s/%s", newMRC.GetNamespace(), newMRC.GetName())
+		}
+
+		if oldMRC.Spec.TrustDomain != newMRC.Spec.TrustDomain {
+			return nil, fmt.Errorf("cannot update trust domain for MRC %s/%s", newMRC.GetNamespace(), newMRC.GetName())
+		}
+
+		if oldMRC.Status.State == newMRC.Status.State {
+			return nil, nil
+		}
+
+		if err := validateMeshRootCertificateStatusTransition(oldMRC, newMRC); err != nil {
+			return nil, err
+		}
+
+		if cv.activeRootCertificateRotation(newMRC) {
+			return nil, fmt.Errorf("cannot update MRC %s/%s root with status %s. Certificate rotation in progress. At most, only 2 MRCs can be in non error or non inactive states",
+				newMRC.GetNamespace(), newMRC.GetName(), newMRC.Status.State)
+		}
+	}
+
+	return nil, nil
+}
+
+// activeRootCertificateRotation returns true if there is an active OSM root certificate rotation, false otherwise
+func (cv *configValidator) activeRootCertificateRotation(newMRC *configv1alpha2.MeshRootCertificate) bool {
+	mrcList, err := cv.configClient.ConfigV1alpha2().MeshRootCertificates(cv.osmNamespace).List(context.TODO(), v1.ListOptions{})
+	if err != nil {
+		log.Error().Err(err)
```

Comment:
The log package requires you call `Msg` or `Msgf` to flush the log. You can also just use `log.Err(err).Msg(...)` without the `.Error().Err(err)`